Add docs for valueType function (#687)
neo4j/docs-cypher: modules/ROOT/pages/values-and-types/property-structural-constructed.adoc

The table below shows the types and their syntactic synonyms.
These types (and their synonyms) can be used in xref::syntax/expressions.adoc#type-predicate-expressions[type predicate expressions] and in xref::constraints/examples.adoc#constraints-examples-node-property-type[node] and xref::constraints/examples.adoc#constraints-examples-relationship-property-type[relationship] property type constraints.

> They are also output as a `STRING` value from the xref::functions/scalar.adoc#functions-valueType[valueType()] function.

Reviewer: "output" is a little bit awkward as a verb. I think the below suggestion is accurate?

```suggestion
They are also returned as a `STRING` value when using the xref::functions/scalar.adoc#functions-valueType[valueType()] function.
```

All Cypher types contain the `null` value. To make them not nullable, `NOT NULL` can be appended to the end of the type (e.g. `BOOLEAN NOT NULL`, `LIST<FLOAT NOT NULL>`).
Note that closed dynamic types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be appended with `NOT NULL`: all inner types must be nullable, or all appended with `NOT NULL`.

== Type Normalization

> Cypher runs a normalization algorithm on all input types, simplifying the given type to the smallest precise representation.

Reviewer 2: I don't think smallest is correct, if you think at the normalization rule: PROPERTY TYPE -> unwrap to individual types.
```suggestion
Cypher runs a normalization algorithm on all input types, simplifying the given type to a deterministic representation for equivalent types.
```

Types are simplified to their default name (e.g. `BOOL` is simplified to `BOOLEAN`), encompassing types are absorbed (e.g. `LIST<BOOLEAN> | LIST<BOOLEAN | INTEGER>` is normalized to `LIST<BOOLEAN | INTEGER>`) and they are ordered.

> The order is based on cardinality, with lower coming first (e.g. `BOOLEAN` comes before `STRING`).

Reviewer 2: I think this is hard to understand, since even when writing the CIP, cardinality feels very fuzzy. I would prefer to just list the order in the documentation and mention that the following holds true:
- subtypes are ordered before their supertypes the resulting order is deterministic
- resulting order is deterministic

Author: I have now updated this to try to cover your suggestion here :) Let me know what you think!

All Cypher types contain the `null` value. To make them not nullable, `NOT NULL` can be appended to the end of the type (e.g. `BOOLEAN NOT NULL`, `LIST<FLOAT NOT NULL>`).
Note that closed dynamic types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be appended with `NOT NULL`: all inner types must be nullable, or all appended with `NOT NULL`.

== Type Normalization

Cypher runs a normalization algorithm on all input types, simplifying the given type to a deterministic representation for equivalent types.

> Types are simplified to their default name (e.g. `BOOL` is simplified to `BOOLEAN`), encompassing types are absorbed (e.g. `LIST<BOOLEAN> | LIST<BOOLEAN | INTEGER>` is normalized to `LIST<BOOLEAN | INTEGER>`) and they are <<ordering-of-types, ordered>>.

Reviewer: xrefs are easier to spot in case we move material and the link breaks.
Also, I think breaking this up into smaller sentences makes it easier to read. 
Do we need to clarify what "encompassing types" are? I think I understand, but not 100% to be honest. 

```suggestion
Types are simplified to their default name (e.g. `BOOL` is simplified to `BOOLEAN`). 
Encompassing types are absorbed (e.g. `LIST<BOOLEAN> | LIST<BOOLEAN | INTEGER>` and normalized to `LIST<BOOLEAN | INTEGER>`).
Types are also xref:values-and-types/property-structural-constructed.adoc#ordering-of-types[ordered].
```

Author: An encompassing type is one that contains all the types of another type, I am happy to add more here but not sure exactly what I'd write :)

Reviewer: I think it's probably fine as is - that was how I understood it without any further descriptions.

The type `PROPERTY VALUE` is expanded to a closed dynamic union of all valid property types, and if all types are represented, then the normalization would simplify to `ANY`.

For example, given the closed dynamic type `BOOL | LIST<INT> | BOOLEAN | LIST<FLOAT | INT>`, the normalized type would be: `BOOLEAN | LIST<INTEGER | FLOAT>`.

> This normalization is run on types input in xref::syntax/expressions.adoc#type-predicate-expressions[type predicate expressions] and in xref::constraints/examples.adoc#constraints-examples-node-property-type[node] and xref::constraints/examples.adoc#constraints-examples-relationship-property-type[relationship] property type constraints.

Reviewer: "input in" reads a bit strange, but maybe "used" is not correct? "given to", perhaps?

```suggestion
This normalization is run on types used in xref::syntax/expressions.adoc#type-predicate-expressions[type predicate expressions], and in xref::constraints/examples.adoc#constraints-examples-node-property-type[node] and xref::constraints/examples.adoc#constraints-examples-relationship-property-type[relationship] property type constraints.
```

> It is also used to ensure the consistency of the output for the xref::functions/scalar.adoc#functions-valueType[valueType()] function.

Reviewer: ```suggestion
Type normalization is also used to ensure the consistency of the output for the xref::functions/scalar.adoc#functions-valueType[valueType()] function.
```

All Cypher types contain the `null` value. To make them not nullable, `NOT NULL` can be appended to the end of the type (e.g. `BOOLEAN NOT NULL`, `LIST<FLOAT NOT NULL>`).
Note that closed dynamic types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be appended with `NOT NULL`: all inner types must be nullable, or all appended with `NOT NULL`.

> [[type-normalization]]]

Reviewer: ```suggestion
[[type-normalization]]
```
